src/main/java/com/google/devtools/build/lib/analysis/IncompatiblePlatformProvider.java

```diff
@@ -0,0 +1,81 @@
+// Copyright 2020 The Bazel Authors. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package com.google.devtools.build.lib.analysis;
+
+import com.google.common.base.Preconditions;
+import com.google.devtools.build.lib.analysis.platform.ConstraintValueInfo;
+import com.google.devtools.build.lib.concurrent.ThreadSafety.Immutable;
+import javax.annotation.Nullable;
+
+/**
+ * Provider instance for the {@code target_compatible_with} attribute.
+ *
+ * <p>The presence of this provider is used to indicate that a target is incompatible with the
+ * current platform. Any target that provides this will automatically be excluded from {@link
+ * SkyframeAnalysisResult}'s list of configured targets.
+ *
+ * <p>This provider is able to keep track of _why_ the corresponding target is considered
+ * incompatible. If the target is incompatible because the target platform didn't satisfy one of the
+ * constraints in target_compatible_with, then the relevant constraint is accessible via {@code
+ * getConstraintResponsibleForIncompatibility()}. On the other hand, if the corresponding target is
+ * incompatible because one of its dependencies is incompatible, then the incompatible dependency is
+ * available via {@code getTargetResponsibleForIncompatibility()}.
+ */
+@Immutable
+public final class IncompatiblePlatformProvider implements TransitiveInfoProvider {
+  private final ConfiguredTarget targetResponsibleForIncompatibility;
```

Comment:
This is my only substantial concern: it's possible a for a target to be incompatible because multiple dependent targets are incompatible, or because multiple constraints in a "target_compatible_with" don't match. Ideally we'd report all of these at once. Otherwise, we report one constraint, the user fixes it, then we report the next constraint, and the user has to fix that separately. Reporting as many errors at once makes the user's task simpler.

Reply:
I agree that would be a better user experience, but I'm a little worried how to report it usefully to the user without making that into another larger project.  There is a point where bazel query and friends are a better tool for finding and rendering graph problems.

Say we have a -> b -> c, and a -> d -> c.  c is incompatible.  If you fix the problem with b, you'll then get d.  How do we display that?  Do we just display one of the paths or all?

The simpler case is a -> b -> c, and a -> d -> e, where c and e are incompatible.  Both dependency chains make sense then to display.

Looking at precedence in Bazel for how this is handled elsewhere, we see that if we build a diamond graph in Bazel with a loop ( a -> b -> d -> a, and a -> c -> d -> a) with cc_library, for example, we get the following error:
```
austin[46124] aschuh-3950x (master) ~/local/code/testing
$ bazel build //testing/a:main 
ERROR: /home/austin/local/code/testing/a/BUILD:1:11: in cc_library rule //testing/a:main: cycle in dependency graph:
.-> //testing/a:main
|   //testing/c:c
|   //testing/d:d
`-- //testing/a:main
This cycle occurred because of a configuration option
ERROR: Analysis of target '//testing/a:main' failed; build aborted
INFO: Elapsed time: 0.973s
INFO: 0 processes.
FAILED: Build did NOT complete successfully (0 packages loaded, 0 targets configured)
```
Fixing the c-> d gives us:
```
austin[46126] aschuh-3950x (master) ~/local/code/testing
$ bazel build //testing/a:main 
ERROR: /home/austin/local/code/testing/a/BUILD:1:11: in cc_library rule //testing/a:main: cycle in dependency graph:
.-> //testing/a:main
|   //testing/b:b
|   //testing/d:d
`-- //testing/a:main
This cycle occurred because of a configuration option
ERROR: Analysis of target '//testing/a:main' failed; build aborted
INFO: Elapsed time: 1.267s
INFO: 0 processes.
FAILED: Build did NOT complete successfully (1 packages loaded, 1 target configured)
```

Doing this with public vs private gives us the same thing:

```
austin[46128] aschuh-3950x (master) ~/local/code/testing
$ bazel build //testing/a:main 
ERROR: /home/austin/local/code/testing/b/BUILD:1:11: in cc_library rule //testing/b:b: target '//testing/d:d' is not visible from target '//testing/b:b'. Check the visibility declaration of the former target if you think the dependency is legitimate
ERROR: Analysis of target '//testing/a:main' failed; build aborted: Analysis failed
INFO: Elapsed time: 5.395s
INFO: 0 processes.
FAILED: Build did NOT complete successfully (2 packages loaded, 3 targets configured)
austin[46129] aschuh-3950x (master) ~/local/code/testing
```
then when the first dependency is removed,
```
austin[46130] aschuh-3950x (master) ~/local/code/testing
$ bazel build //testing/a:main 
ERROR: /home/austin/local/code/testing/c/BUILD:2:11: in cc_library rule //testing/c:c: target '//testing/d:d' is not visible from target '//testing/c:c'. Check the visibility declaration of the former target if you think the dependency is legitimate
ERROR: Analysis of target '//testing/a:main' failed; build aborted: Analysis of target '//testing/c:c' failed
INFO: Elapsed time: 3.884s
INFO: 0 processes.
FAILED: Build did NOT complete successfully (1 packages loaded, 1 target configured)
```

I like reporting multiple constraints.  That's easy and high value.  Would it work to tackle that problem now, and address showing more errors driven by user feedback?

Reply:
I added tracking since it was easy.  I didn't touch displaying that information though, choosing to instead just show the first chain.  I'd be happy to display it if someone has a suggestion on a good way to render it.

Reply:
I am fine to collect the information, and just report a subset with a TODO to improve the error reporting.

Reply:
TODO added.